torch/lib/ATen/DLConvertor.cpp

```diff
@@ -125,28 +125,41 @@ static ScalarType getATenScalarType(const DLDataType& dtype) {
 }
 
 
-// This function returns a shared_ptr to DLpack tensor constructed out ATen tensor
-DLTensor* toDLPack(const Tensor& src, DLTensor* dlTensor) {
-  dlTensor->data = src.data_ptr();
+void destructor(void * ctx, void * arg) {
+  static_cast<TensorImpl*>(ctx)->release();
+  delete static_cast<DLManagedTensor*>(arg);
+}
+
+
+// This function returns a shared_ptr to DLpack tensor constructed out of ATen tensor
+DLManagedTensor* toDLPack(const Tensor& src) {
+  // retain the underlying storage to make sure it doesn't go out of scope
+  src.get()->retain();
+  // since we manage memory now, we can do this allocation here
+  DLManagedTensor* dlMTensor(new DLManagedTensor);
```

Comment:
nit: allocate this `dlTensor` before you call `retain()` (e.g. using unique_ptr), or `release()` in case of an error (this is an uglier solution).